Foundation/JSONSerialization.swift

```diff
@@ -285,219 +454,176 @@ internal extension JSONSerialization {
                 case "\u{10}"..."\u{1f}":
-                    writer("\\u00\(String(scalar.value, radix: 16))") // U+0010 to U+001F
+                    result += "\\u00\(String(scalar.value, radix: 16))" // U+0010 to U+001F
                 default:
-                    writer(String(scalar))
+                    result += String(scalar)
             }
         }
-        writer("\"")
+        result += "\""
     }
 
-    private func serializeFloat<T: FloatingPoint & LosslessStringConvertible>(_ num: T) throws {
-        guard num.isFinite else {
-             throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListReadCorrupt.rawValue, userInfo: ["NSDebugDescription" : "Invalid number value (\(num)) in JSON write"])
-        }
-        var str = num.description
+    mutating func visit<F: FloatingPoint & LosslessStringConvertible>(_ num: F) throws {
+        var str = num.description //TODO
         if str.hasSuffix(".0") {
             str.removeLast(2)
         }
-        writer(str)
+        result += str
     }
-
-    mutating func serializeNumber(_ num: NSNumber) throws {
-        if CFNumberIsFloatType(num._cfObject) {
-            try serializeFloat(num.doubleValue)
-        } else {
-            switch num._cfTypeID {
-            case CFBooleanGetTypeID():
-                writer(num.boolValue.description)
-            default:
-                writer(num.stringValue)
-            }
-        }
-    }
-
-    mutating func serializeArray(_ array: [Any?]) throws {
-        writer("[")
+    
+    mutating func visitArrayLikeSequence<Seq: Sequence>(_ array: Seq) throws {
+        result += "["
         if pretty {
-            writer("\n")
+            result += "\n"
             incAndWriteIndent()
         }
         
         var first = true
         for elem in array {
+            guard let jsonElem = elem as? _JSONValue else {
+                throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListReadCorrupt.rawValue, userInfo: ["NSDebugDescription" : "Invalid NSArray value (\(elem)) in JSON write"])
+            }
             if first {
                 first = false
             } else if pretty {
-                writer(",\n")
+                result += ",\n"
                 writeIndent()
             } else {
-                writer(",")
+                result += ","
             }
-            try serializeJSON(elem)
+            try jsonElem.visit(with: &self)
         }
         if pretty {
-            writer("\n")
+            result += "\n"
             decAndWriteIndent()
         }
-        writer("]")
+        result += "]"
     }
 
-    mutating func serializeDictionary(_ dict: Dictionary<AnyHashable, Any?>) throws {
-        writer("{")
+    mutating func visit<J: _JSONValue>(_ array: [J]) throws {
+        try visitArrayLikeSequence(array)
+    }
+    
+    mutating func visit(_ cocoaArray: NSArray) throws {
+        try visitArrayLikeSequence(cocoaArray)
```

Comment:
Should do a visitDictionaryLikeSequence one too to avoid bridging NSDictionary just to write it out